presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -0,0 +1,280 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.analyzer;
+
+import com.facebook.presto.sql.tree.ArithmeticBinaryExpression;
+import com.facebook.presto.sql.tree.AstVisitor;
+import com.facebook.presto.sql.tree.ComparisonExpression;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.GroupBy;
+import com.facebook.presto.sql.tree.GroupingElement;
+import com.facebook.presto.sql.tree.Identifier;
+import com.facebook.presto.sql.tree.LogicalBinaryExpression;
+import com.facebook.presto.sql.tree.Node;
+import com.facebook.presto.sql.tree.OrderBy;
+import com.facebook.presto.sql.tree.Query;
+import com.facebook.presto.sql.tree.QueryBody;
+import com.facebook.presto.sql.tree.QuerySpecification;
+import com.facebook.presto.sql.tree.Relation;
+import com.facebook.presto.sql.tree.Select;
+import com.facebook.presto.sql.tree.SelectItem;
+import com.facebook.presto.sql.tree.SimpleGroupBy;
+import com.facebook.presto.sql.tree.SingleColumn;
+import com.facebook.presto.sql.tree.SortItem;
+import com.facebook.presto.sql.tree.Table;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Optional;
+
+import static com.facebook.presto.sql.analyzer.MaterializedViewInformationExtractor.MaterializedViewInfo;
+import static java.util.Objects.requireNonNull;
+
+public class MaterializedViewQueryOptimizer
+        extends AstVisitor<Node, Void>
+{
+    private final Table materializedView;
+
+    private final MaterializedViewInfo materializedViewInfo;
+
+    private boolean isRewritable = true;
+
+    public MaterializedViewQueryOptimizer(Table materializedView, Query materializedViewQuery)
+    {
+        this.materializedView = requireNonNull(materializedView, "materialized view is null");
+        MaterializedViewInfo materializedViewInfo = new MaterializedViewInfo();
+        MaterializedViewInformationExtractor materializedViewInformationExtractor = new MaterializedViewInformationExtractor();
+        materializedViewInformationExtractor.process(materializedViewQuery, materializedViewInfo);
+        this.materializedViewInfo = materializedViewInfo;
+
+        // If materialized view has a LIMIT clause, then it would reject the rewrite optimization
+        if (materializedViewInfo.isLimitClausePresented()) {
```

Comment:
If we throw exception from the extractor then we don't need to check for the limit clause here.